# Price Cut Los Angeles Random Forest

### Data Cleanup

The following property types were eliminated from the original CSV:
1. Lot/Land

In [1]:
# Dependencies
import pandas as pd
from pathlib import Path

In [2]:
# Name of the CSV file
file_path = Path("Resources/Zillow_Los_Angeles_For_Sale.csv")
df = pd.read_csv(file_path)
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/406-N-Oakhu...     NaN   
1    https://www.zillow.com/homedetails/460-N-Palm-...     NaN   
2    https://www.zillow.com/homedetails/321-N-Oakhu...     NaN   
3    https://www.zillow.com/homedetails/9233-Burton...     NaN   
4    https://www.zillow.com/homedetails/450-N-Palm-...     NaN   
..                                                 ...     ...   
154  https://www.zillow.com/homedetails/9814-Curwoo...     NaN   
155  https://www.zillow.com/homedetails/509-N-Elm-D...     NaN   
156  https://www.zillow.com/homedetails/9153-Janice...     NaN   
157  https://www.zillow.com/homedetails/1530-N-Beve...     NaN   
158  https://www.zillow.com/homedetails/9669-Oak-Pa...     NaN   

    Listing description Property type  Number of days on Zillow  \
0        Condo for sale   Condominium                       NaN   
1        Condo for sale   Condominium                       NaN   
2        Condo for sale   Condominium                      16.0   
3        Condo for sale   Condominium                       NaN   
4        Condo for sale   Condominium                       NaN   
..                  ...           ...                       ...   
154    New construction       Unknown                      45.0   
155    New construction       Unknown                      35.0   
156    New construction       Unknown                       NaN   
157    New construction       Unknown                      43.0   
158    New construction       Unknown                     178.0   

     Property price (USD) Price was cut  Price cut amount (USD)  \
0                 1295000            No                     NaN   
1                 4495000            No                     NaN   
2                  995000            No                     NaN   
3                 1345000           Yes                 50000.0   
4                 3980000            No                     NaN   
..                    ...           ...                     ...   
154              12500000            No                     NaN   
155              15950000            No                     NaN   
156              15495000            No                     NaN   
157               6850000            No                     NaN   
158              29950000            No                     NaN   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                           NaN       1293.0  ...      2.0        2.0   
1                           NaN       2882.0  ...      4.0        4.0   
2                           NaN       1791.0  ...      2.0        2.0   
3                        2/3/24       1567.0  ...      2.0        3.0   
4                           NaN       2958.0  ...      3.0        4.0   
..                          ...          ...  ...      ...        ...   
154                         NaN       9500.0  ...      6.0        9.0   
155                         NaN       9932.0  ...      6.0        9.0   
156                         NaN       6194.0  ...      6.0        9.0   
157                         NaN       4900.0  ...      6.0        8.0   
158                         NaN      16900.0  ...      8.0       13.0   

                      Broker name                      Broker agent  \
0                    Rodeo Realty      Sandy Richman DRE # 01239013   
1                      The Agency      Azy Farahmand DRE # 01958066   
2            Sync Brokerage, Inc.        Alicia Shen DRE # 02026588   
3                  Diamond Realty                               NaN   
4                  Christie's AKG   Nicole Contreras DRE # 01512844   
..                            ...                               ...   
154       Rodeo Realty- Brentwood  Haleh Dowlatshahi DRE # 01455468   
155                    The Agency      Azy Farahmand DRE # 01958066   
156  Estates & Trusts Realty, Inc          Jody Katz DRE # 02013896   
157     The Beverly Hills Estates          Ziv Gabay

In [3]:
# Clean the dataframe by replacing NaN values with 0
df.fillna(0, inplace=True)

In [4]:
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/406-N-Oakhu...     0.0   
1    https://www.zillow.com/homedetails/460-N-Palm-...     0.0   
2    https://www.zillow.com/homedetails/321-N-Oakhu...     0.0   
3    https://www.zillow.com/homedetails/9233-Burton...     0.0   
4    https://www.zillow.com/homedetails/450-N-Palm-...     0.0   
..                                                 ...     ...   
154  https://www.zillow.com/homedetails/9814-Curwoo...     0.0   
155  https://www.zillow.com/homedetails/509-N-Elm-D...     0.0   
156  https://www.zillow.com/homedetails/9153-Janice...     0.0   
157  https://www.zillow.com/homedetails/1530-N-Beve...     0.0   
158  https://www.zillow.com/homedetails/9669-Oak-Pa...     0.0   

    Listing description Property type  Number of days on Zillow  \
0        Condo for sale   Condominium                       0.0   
1        Condo for sale   Condominium                       0.0   
2        Condo for sale   Condominium                      16.0   
3        Condo for sale   Condominium                       0.0   
4        Condo for sale   Condominium                       0.0   
..                  ...           ...                       ...   
154    New construction       Unknown                      45.0   
155    New construction       Unknown                      35.0   
156    New construction       Unknown                       0.0   
157    New construction       Unknown                      43.0   
158    New construction       Unknown                     178.0   

     Property price (USD) Price was cut  Price cut amount (USD)  \
0                 1295000            No                     0.0   
1                 4495000            No                     0.0   
2                  995000            No                     0.0   
3                 1345000           Yes                 50000.0   
4                 3980000            No                     0.0   
..                    ...           ...                     ...   
154              12500000            No                     0.0   
155              15950000            No                     0.0   
156              15495000            No                     0.0   
157               6850000            No                     0.0   
158              29950000            No                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                             0       1293.0  ...      2.0        2.0   
1                             0       2882.0  ...      4.0        4.0   
2                             0       1791.0  ...      2.0        2.0   
3                        2/3/24       1567.0  ...      2.0        3.0   
4                             0       2958.0  ...      3.0        4.0   
..                          ...          ...  ...      ...        ...   
154                           0       9500.0  ...      6.0        9.0   
155                           0       9932.0  ...      6.0        9.0   
156                           0       6194.0  ...      6.0        9.0   
157                           0       4900.0  ...      6.0        8.0   
158                           0      16900.0  ...      8.0       13.0   

                      Broker name                      Broker agent  \
0                    Rodeo Realty      Sandy Richman DRE # 01239013   
1                      The Agency      Azy Farahmand DRE # 01958066   
2            Sync Brokerage, Inc.        Alicia Shen DRE # 02026588   
3                  Diamond Realty                                 0   
4                  Christie's AKG   Nicole Contreras DRE # 01512844   
..                            ...                               ...   
154       Rodeo Realty- Brentwood  Haleh Dowlatshahi DRE # 01455468   
155                    The Agency      Azy Farahmand DRE # 01958066   
156  Estates & Trusts Realty, Inc          Jody Katz DRE # 02013896   
157     The Beverly Hills Estates          Ziv Gabay

In [5]:
# Import scikit-learn libraries including Random Forest Classifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier  # Changed from DecisionTreeClassifier to RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [6]:
# Convert 'Price was cut' to a binary indicator (0 and 1)
df['Price was cut'] = df['Price was cut'].apply(lambda x: 1 if x == 'Yes' else 0)
df.tail()

Property URL  MLS ID  \
154  https://www.zillow.com/homedetails/9814-Curwoo...     0.0   
155  https://www.zillow.com/homedetails/509-N-Elm-D...     0.0   
156  https://www.zillow.com/homedetails/9153-Janice...     0.0   
157  https://www.zillow.com/homedetails/1530-N-Beve...     0.0   
158  https://www.zillow.com/homedetails/9669-Oak-Pa...     0.0   

    Listing description Property type  Number of days on Zillow  \
154    New construction       Unknown                      45.0   
155    New construction       Unknown                      35.0   
156    New construction       Unknown                       0.0   
157    New construction       Unknown                      43.0   
158    New construction       Unknown                     178.0   

     Property price (USD)  Price was cut  Price cut amount (USD)  \
154              12500000              0                     0.0   
155              15950000              0                     0.0   
156              15495000              0                     0.0   
157               6850000              0                     0.0   
158              29950000              0                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
154                           0       9500.0  ...      6.0        9.0   
155                           0       9932.0  ...      6.0        9.0   
156                           0       6194.0  ...      6.0        9.0   
157                           0       4900.0  ...      6.0        8.0   
158                           0      16900.0  ...      8.0       13.0   

                      Broker name                      Broker agent  \
154       Rodeo Realty- Brentwood  Haleh Dowlatshahi DRE # 01455468   
155                    The Agency      Azy Farahmand DRE # 01958066   
156  Estates & Trusts Realty, Inc          Jody Katz DRE # 02013896   
157     The Beverly Hills Estates          Ziv Gabay DRE # 01971047   
158         Hurwitz James Company        Bob Hurwitz DRE # 00526195   

        Street address                       City    Zip State Country  \
154    9814 Curwood Pl              Beverly Hills  90210    CA     USA   
155       509 N Elm Dr              Beverly Hills  90210    CA     USA   
156     9153 Janice Pl              Beverly Hills  90210    CA     USA   
157  1530 N Beverly Dr  Beverly Hills Post Office  90210    CA     USA   
158   9669 Oak Pass Rd              Beverly Hills  90210    CA     USA   

    FederalInt  
154       5.53  
155       5.53  
156       5.53  
157       5.53  
158       5.53  

[5 rows x 24 columns]

In [7]:
# Feature selection
# Selecting relevant features and the target variable
X = df[['Property price (USD)', 'Property type', 'Living area', 'Lot/land area', 'Bedrooms', 'Bathrooms', 'City', 'FederalInt']]
y = df['Price was cut']


In [8]:
# Handle the categorical variables
categorical_features = ['Property type', 'City']
label_encoders = {}

for col in categorical_features:
    le = LabelEncoder()
    X.loc[:,col] = le.fit_transform(X[col])
    label_encoders[col] = le

In [9]:
df

Property URL  MLS ID  \
0    https://www.zillow.com/homedetails/406-N-Oakhu...     0.0   
1    https://www.zillow.com/homedetails/460-N-Palm-...     0.0   
2    https://www.zillow.com/homedetails/321-N-Oakhu...     0.0   
3    https://www.zillow.com/homedetails/9233-Burton...     0.0   
4    https://www.zillow.com/homedetails/450-N-Palm-...     0.0   
..                                                 ...     ...   
154  https://www.zillow.com/homedetails/9814-Curwoo...     0.0   
155  https://www.zillow.com/homedetails/509-N-Elm-D...     0.0   
156  https://www.zillow.com/homedetails/9153-Janice...     0.0   
157  https://www.zillow.com/homedetails/1530-N-Beve...     0.0   
158  https://www.zillow.com/homedetails/9669-Oak-Pa...     0.0   

    Listing description Property type  Number of days on Zillow  \
0        Condo for sale   Condominium                       0.0   
1        Condo for sale   Condominium                       0.0   
2        Condo for sale   Condominium                      16.0   
3        Condo for sale   Condominium                       0.0   
4        Condo for sale   Condominium                       0.0   
..                  ...           ...                       ...   
154    New construction       Unknown                      45.0   
155    New construction       Unknown                      35.0   
156    New construction       Unknown                       0.0   
157    New construction       Unknown                      43.0   
158    New construction       Unknown                     178.0   

     Property price (USD)  Price was cut  Price cut amount (USD)  \
0                 1295000              0                     0.0   
1                 4495000              0                     0.0   
2                  995000              0                     0.0   
3                 1345000              1                 50000.0   
4                 3980000              0                     0.0   
..                    ...            ...                     ...   
154              12500000              0                     0.0   
155              15950000              0                     0.0   
156              15495000              0                     0.0   
157               6850000              0                     0.0   
158              29950000              0                     0.0   

    Price cut date (MM/DD/YYYY)  Living area  ... Bedrooms  Bathrooms  \
0                             0       1293.0  ...      2.0        2.0   
1                             0       2882.0  ...      4.0        4.0   
2                             0       1791.0  ...      2.0        2.0   
3                        2/3/24       1567.0  ...      2.0        3.0   
4                             0       2958.0  ...      3.0        4.0   
..                          ...          ...  ...      ...        ...   
154                           0       9500.0  ...      6.0        9.0   
155                           0       9932.0  ...      6.0        9.0   
156                           0       6194.0  ...      6.0        9.0   
157                           0       4900.0  ...      6.0        8.0   
158                           0      16900.0  ...      8.0       13.0   

                      Broker name                      Broker agent  \
0                    Rodeo Realty      Sandy Richman DRE # 01239013   
1                      The Agency      Azy Farahmand DRE # 01958066   
2            Sync Brokerage, Inc.        Alicia Shen DRE # 02026588   
3                  Diamond Realty                                 0   
4                  Christie's AKG   Nicole Contreras DRE # 01512844   
..                            ...                               ...   
154       Rodeo Realty- Brentwood  Haleh Dowlatshahi DRE # 01455468   
155                    The Agency      Azy Farahmand DRE # 01958066   
156  Estates & Trusts Realty, Inc          Jody Katz DRE # 02013896   
157     The Beverly Hills Estates       

In [10]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [11]:
# Initializing and training the Random Forest Classifier
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [12]:
# Make predictions
y_pred = clf.predict(X_test)


In [13]:
# Evaluating the model
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy Score: 0.625
Confusion Matrix:
 [[20  6]
 [ 6  0]]
Classification Report:
               precision    recall  f1-score   support

           0       0.77      0.77      0.77        26
           1       0.00      0.00      0.00         6

    accuracy                           0.62        32
   macro avg       0.38      0.38      0.38        32
weighted avg       0.62      0.62      0.62        32



# Initial Result of the Decision Tree Model

### Accuracy Score
The accuracy score of 0.625 (or 62.5%) indicates that the model correctly predicts the outcome (whether the price will be cut or not) in approximately 62.5% of the cases. This metric provides a general idea of the model's overall performance but doesn't detail its effectiveness in distinguishing between the two classes.


In [14]:
# Note: Random Forest trees might be too complex for practical visualization with plot_tree,
# especially for large numbers of trees. Consider summarizing tree characteristics or feature importances instead.
feature_importances = pd.Series(clf.feature_importances_, index=X.columns).sort_values(ascending=False)
print(feature_importances)

Property price (USD)    0.267348
Living area             0.225895
Lot/land area           0.221811
Bathrooms               0.123934
Bedrooms                0.118128
Property type           0.041705
City                    0.001178
FederalInt              0.000000
dtype: float64
